src/apm/listener.rs

```diff
@@ -0,0 +1,60 @@
+use std::ops::{Deref, DerefMut};
+use std::sync::RwLock;
+
+use apm::event::{CommandStarted, CommandResult};
+use error::{Error, Result};
+
+pub struct Listener {
+    start_hooks: RwLock<Vec<fn(&CommandStarted)>>,
+    completion_hooks: RwLock<Vec<fn(&CommandResult)>>,
+}
+
+impl Listener {
+    pub fn new() -> Listener {
+        Listener { start_hooks: RwLock::new(vec![]), completion_hooks: RwLock::new(vec![]) }
+    }
+
+    pub fn add_start_hook(&self, hook: fn(&CommandStarted)) -> Result<()> {
+        let mut guard = match self.start_hooks.write() {
+            Ok(guard) => guard,
+            Err(_) => return Err(Error::PoisonLockError)
+        };
+
+        Ok(guard.deref_mut().push(hook))
+    }
+
+    pub fn add_completion_hook(&self, hook: fn(&CommandResult)) -> Result<()> {
+        let mut guard = match self.completion_hooks.write() {
+            Ok(guard) => guard,
+            Err(_) => return Err(Error::PoisonLockError)
+        };
+
+        Ok(guard.deref_mut().push(hook))
+    }
+
+    pub fn run_start_hooks(&self, started: &CommandStarted) -> Result<()> {
```

Comment:
Hmmm...looks like this might actually not be feasible. Since `Client` is just an `Arc`, it's fields are enforced as immutable, hence the need for the mutex on the arrays in the first place. In order to change the boolean, it would have to *also* be wrapped in a mutex, which kind of defeats the purpose :/